src/switcher.S

```diff
@@ -0,0 +1,134 @@
+#include "comps_offsets.h"
+
+.text
+.balign 4
+
+.global switcher_entry
+.global switch_compartment
+.global switch_compartment_end
+
+/**
+ * Entry point from user code to switcher function
+ *
+ * @param c0 DDC of switcher, containing compartment information
+ */
+.type switcher_entry, "function"
+switcher_entry:
+    mov      c29, c0
+    mov      x0, #0
+    cvtp     clr, lr
+    b        switch_compartment
+    ret      clr
+
+/** Code to perform actual switch
+ *
+ * @param x0 ID (in `comps` array) of compartment to switch to
+ */
+.type switch_compartment, "function"
+switch_compartment:
+    // Store entering compartment's DDC, and move to memory containing
+    // compartment info
+    mrs       c2, DDC
+    mov       x10, x0
+
+    // Expect switcher DDC in c29
+    msr       DDC, c29
+
+    // Get compartment to switch to data
+    mov       x11, #COMP_SIZE
+    mul       x10, x10, x11
+
+    // Load PCC, including function we are jumping to within compartment
+    add       x11, x10, #COMP_OFFSET_PCC
+    ldr       c0, [x29, x11]
+
+    // Load DDC
+    add       x11, x10, #COMP_OFFSET_DDC
+    ldr       c1, [x29, x11]
+
+    // Setup SP
+    mov       x12, sp
+    add       x11, x10, #COMP_OFFSET_STK_ADDR
+    ldr       x11, [x29, x11]
+    mov       sp, x11
+
+    // Install compartment DDC
+    msr       DDC, c1
+
+    // Save old DDC (c2), old SP (x12), old CLR (clr) on stack
+    stp       c2, clr, [sp, #-48]!
+    str       x12, [sp, #32]
```

Comment:
Doesn't this make them reachable from the callee compartment? If so, why isn't it a security hole?

Reply:
Aha, we found a security hole in what we considered a secure implementation of switching [1]. Indeed, these should be saved before we swap the DDC. I will keep this in mind as I work on the switcher. We could probably add a test which shows this can be done on the prior example, then fix the security hole.

[1] https://github.com/capablevms/cheri-examples/blob/master/hybrid/compartment_examples/inter_comp_call/secure/switch_compartment.s#L35-L40